index.js

```diff
@@ -138,82 +89,170 @@ ModeDevice.prototype.setUpApiHttp = function() {
+ModeDevice.prototype._wsHandlePong = function(msg) {
+  debuglog('Received websocket pong: %s', msg);
+  if (msg == this._wsPendingPingMsg) {
+    this._wsPendingPingMsg = null;
   }
+};
 
-  var proto = this.apiUseTls ? 'wss' : 'ws';
-  var target = proto + '://' + this.apiHost + ':' + this.apiPort + '/devices/' + this.deviceId + '/command';
+ModeDevice.prototype._wsConnect = function() {
+  const proto = this.apiUseTls ? 'wss' : 'ws';
+  const target = proto + '://' + this.apiHost + ':' + this.apiPort + '/devices/' + this.deviceId + '/command';
 
-  debuglog("Connecting to " + target);
-  this.websocket = new ws(target, {
+  debuglog('Making websocket connection to %s', target);
+  this._websocket = new ws(target, {
     agent: this.apiHttpAgent,
     headers: {
       "Authorization": 'ModeCloud ' + this.token
     }
   });
-  this.websocket.on('error', this.errorCallback.bind(this));
-  this.websocket.on('close', this.closeCallback.bind(this));
-  this.websocket.on('open', this.openCallback.bind(this));
-  this.websocket.on('message', commandHandler.bind(this));
-  this.websocket.on('pong', this.pongCallback.bind(this));
+
+  this._websocket.on('error', this._wsHandleError.bind(this));
+  this._websocket.on('close', this._wsHandleClose.bind(this));
+  this._websocket.on('open', this._wsHandleOpen.bind(this));
+  this._websocket.on('message', this._wsHandleMessage.bind(this));
+  this._websocket.on('pong', this._wsHandlePong.bind(this));
+
+  this._wsReconnectTimer = null;
 };
 
-ModeDevice.prototype.scheduleReconnect = function(firstConnect) {
-  var wait = firstConnect ? 0 : this.retryWait;
-  debuglog('Reconnect websocket in ' + wait + ' seconds');
-  var device = this;
-  if (device.isReconnectScheduled !== true) {
-    debuglog('scheduling a reconnection');
-    device.isReconnectScheduled = true;
-    setTimeout(function() {
-      debuglog('scheduled reconnection triggered');
-      device.reconnect();
-      device.isReconnectScheduled = false;
-    }, wait * 1000);
+ModeDevice.prototype._wsScheduleReconnect = function() {
+  if (!this._listeningCommands) {
+    // This is triggered by a graceful shutdown, so no need to reconnect.
+    return;
   }
 
-  // Only if less than 60 sec, we increment waiting time according to Fibonacci numbers.
-  if (this.retryWait < 60) {
-    var fib = this.retryWaitFib;
-    this.retryWaitFib = this.retryWait;
-    this.retryWait += fib;
+  if (this._wsReconnectDelay < 60) {
+    // exponential backoff
+    this._wsReconnectDelay = Math.pow(2, this._wsReconnectAttempts);
   }
+
+  this._wsReconnectAttempts++;
+  debuglog('Retrying websocket connection in %d seconds (attempt #%d)', this._wsReconnectDelay, this._wsReconnectAttempts);
+
+  // Make sure we never double-schedule reconnection.
+  if (this._wsReconnectTimer) {
+    clearTimeout(this._wsReconnectTimer);
+  }
+
+  this._wsReconnectTimer = setTimeout(this._wsConnect.bind(this), this._wsReconnectDelay * 1000);
```

Comment:
Actually, in both cases, the problem was caused by the fact that when a ping fails, we were calling close() *and* reconnect(). Then the onClose handler will also try to call reconnect(). And we ended up with two "threads" competing with each other.